### PR TITLE
Correct way to extract globals in TGlobalMappedFunction

### PR DESCRIPTION
**Pull request by linev**
In TGlobalMappedFunction all global functions casted to the type: `void *(*)()`
But in reality all used functions have other signature: `void *&(*)()`
To silent compiler warnings intermediate casting to (void*) was used - this is **dangerous** and in generally **wrong**

Patch introduces functor inside TGlobalMappedFunction.
Within lambda each function called as it is and just return value casted to `void *`.

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
In all actual ABIs, pointers to non-member function are simply addresses, a raw pointer. The return type doesn't matter for the storage. For type-safety, correcting the function's return type to be `void* &` would be indeed an improvement.

The cost and complexity of going through lambdas seems fairly high to me. Let's see what @pcanal thinks about this.

**Comment by linev** [comment]
> The cost and complexity of going through lambdas seems fairly high to me. Let's see what @pcanal thinks about this.

Seems to be, globals used very rarely through this list. At least, I did not found any macro there it would happen. May be only in python binding.

Also question to @pcanal - with address is TGlobal::GetAddress() has to return? May be it should be address on global variable like gROOT - not the value of variable. In this case wrong casting was **intentionally wrong**. I will push extra commit to check this.

**Comment by linev** [comment]
It is definitely something wrong.
One cannot cast to same function type following functions:

    TROOT *GetROOT();
    TVirtualPad *&Pad();

Either one or another will return wrong result.

**Comment by linev** [comment]
Just very simple macro:

```
void *getGlobal(const char *name) 
{
   auto gl = gROOT->GetListOfGlobals()->FindObject(name);
   if (!gl) return nullptr;   
   TGlobal *global = dynamic_cast<TGlobal *> (gl);
    return global ? global->GetAddress() : nullptr;
}

void test() 
{
   TCanvas c1;
   printf("gROOT %p %p\n", gROOT, getGlobal("gROOT"));
   printf("gPad %p %p\n", gPad, getGlobal("gPad"));
}
```
In current ROOT master - gROOT prints identical code, gPad is **DIFFERENT**
This the provided patch gPad and gROOT printouts are identical.

**Comment by linev** [comment]
My current finding - function cast was done probably **intentionally** wrong.
It was working while some global functions like gPad has signature:

    TVirtualPad *&Pad()

And when return values was casted to (void *) one got address of variable. 
But code was not work for other several methods:
* ROOT::GetROOT
* TInterpreter::Instance
* TGClient::Instance
One could fix problem with gROOT, but not gInterpreter and gClient.

**Comment by pcanal** [comment]
It looks like you right ...  [Note: gPad's function has a different return type because it is the only one that is modifiable] [hummm ... gDirectory also should be working in the current state]

> One could fix problem with gROOT, but not gInterpreter and gClient.

One could take your idea of the functor, give it  some storage and return the address of that storage.

**Comment by linev** [comment]
> One could take your idea of the functor, give it some storage and return the address of that storage.

There is easy solution with TGClient::Instance - just change return type.

With TInterpreter::Instance it is more complex - there is special workaround for win32. 
But seems to be, workaround no longer required in ROOT6. If @bellenot can remove old code, 
we can also easily change return type of  TInterpreter::Instance.

**Comment by pcanal** [comment]
> There is easy solution with TGClient::Instance - just change return type.

True as long as it is change to a const reference ...

**Comment by Axel-Naumann** [comment]
Can we simply add `const TROOT *&GetROOTPtrRef()`, in the source file of where `gROOT` is injected in the list of globals?

**Comment by linev** [comment]
> Can we simply add const TROOT *&GetROOTPtrRef(), in the source file of where gROOT is injected in the list of globals?

Question - should this function be public? 
And that about other gInterpreter and gClient?

**Comment by linev** [comment]
I hope that now solution is better.
I decide not to change signature of existing static methods like `TGClient::Instance()`, 
but just register globals at the place where real variable can be accessed.

The only assumption - `gPtr2Interpreter` remains nullptr. It is now a case. Function was used on Windows with ROOT5; now it is deactivated. If Bertrand @bellenot decides, that it is needed one need to check once again code in TInterpreter.cxx

File: core/meta/inc/TGlobal.h [hunk 1]
```diff
@@ -48,36 +49,53 @@ class TGlobal : public TDictionary {
...
+// Class to map the "funcky" globals and be able to add them to the list of globals.
+class TGlobalMappedFunction : public TGlobal {
+public:
+   typedef void *(*GlobalFunc_t)();
+   typedef std::function<void *()> GlobalFunctor_t;
+   TGlobalMappedFunction(const char *name, const char *type, GlobalFunc_t funcPtr) : fFuncPtr(funcPtr)
+   {
+      SetNameTitle(name, type);
+   }
+   virtual ~TGlobalMappedFunction() = default;
```
Comment (pcanal) [inline core/meta/inc/TGlobal.h]: Please keep the original alignment of function name :)

File: core/meta/inc/TGlobal.h [hunk 2]
```diff
@@ -48,36 +49,53 @@ class TGlobal : public TDictionary {
...
+      glob->fFunctor = functor;
+      Add(glob);
+   }
+
+private:
+   GlobalFunc_t fFuncPtr{nullptr}; // Function to call to get the address
+
+   GlobalFunctor_t fFunctor; // functor which correctly returns pointer
+
+   TGlobalMappedFunction &operator=(const TGlobal &); // not implemented.
```
Comment (pcanal) [inline core/meta/inc/TGlobal.h]: Use =delete to make it as not implemented.

File: core/meta/inc/TGlobal.h (hunk 1 above)
Comment (linev) [reply core/meta/inc/TGlobal.h]: I use here clang-format. Should I apply hand-made formatting?

**Comment by linev** [comment]
@Axel-Naumann 
Should I merge it? 
That about 6.16 branch?

File: core/meta/inc/TGlobal.h (hunk 1 above)
Comment (pcanal) [reply core/meta/inc/TGlobal.h]: yes for that part. clang-format does not support this kind of alignment.

**Pull request merged**